src/main/java/io/github/rhkiswani/javaff/httpclient/ApacheHttpClient2.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright 2016 Mohamed Kiswani.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.github.rhkiswani.javaff.httpclient;
+
+import io.github.rhkiswani.javaff.exceptions.SmartException;
+import io.github.rhkiswani.javaff.httpclient.exceptions.HttpClientException;
+import io.github.rhkiswani.javaff.httpclient.util.IOUtil;
+import io.github.rhkiswani.javaff.json.JsonHandlerFactory;
+import io.github.rhkiswani.javaff.reflection.ReflectionUtil;
+import org.apache.http.Header;
+import org.apache.http.HttpResponse;
+import org.apache.http.HttpStatus;
+import org.apache.http.NameValuePair;
+import org.apache.http.client.entity.UrlEncodedFormEntity;
+import org.apache.http.client.methods.*;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.http.message.BasicNameValuePair;
+
+import java.io.UnsupportedEncodingException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * @author Mohamed Kiswani
+ * @since 0.0.20
+ *
+ */
+public class ApacheHttpClient2 implements HttpClient{
+
+    @Override
+    public String postJson(String url, String json, Map<String, String> headers) throws HttpClientException {
+        return doJsonRequest(new HttpPost(url), json, headers);
+    }
+
+    @Override
+    public String putJson(String url, String json, Map<String, String> headers) throws HttpClientException {
+        return doJsonRequest(new HttpPut(url), json, headers);
+    }
+
+    private String doJsonRequest(HttpRequestBase method, String json, Map<String, String> headers) {
+        try {
+            if (json == null){
+                throw new HttpClientException(SmartException.NULL_VAL, "Json Object");
+            }
+            CloseableHttpClient c =  HttpClientBuilder.create().build();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *c*. Currently, there are execution paths that do not contain closure statements. Either a) close *c* in a try-finally block or b) close the resource by declaring *c* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).